zookeeper-server/src/main/java/org/apache/zookeeper/server/NettyServerCnxn.java

```diff
@@ -0,0 +1,658 @@
+        if (!factory.cnxns.remove(this)) {
+            LOG.debug("cnxns size:{}", factory.cnxns.size());
+            if (channel.isOpen()) {
+                channel.close();
+            }
+            return;
+        }
+
+        LOG.debug("close in progress for session id: 0x{}", Long.toHexString(sessionId));
+
+        factory.removeCnxnFromSessionMap(this);
+
+        factory.removeCnxnFromIpMap(this, ((InetSocketAddress) channel.remoteAddress()).getAddress());
+
+        if (zkServer != null) {
+            zkServer.removeCnxn(this);
+        }
+
+        if (channel.isOpen()) {
+            // Since we don't check on the futures created by write calls to the channel complete we need to make sure
+            // that all writes have been completed before closing the channel or we risk data loss
+            // See: http://lists.jboss.org/pipermail/netty-users/2009-August/001122.html
+            channel.writeAndFlush(Unpooled.EMPTY_BUFFER).addListener(new ChannelFutureListener() {
+                @Override
+                public void operationComplete(ChannelFuture future) {
+                    future.channel().close().addListener(f -> releaseQueuedBuffer());
+                }
+            });
+        } else {
+            ServerMetrics.getMetrics().CONNECTION_DROP_COUNT.add(1);
+            channel.eventLoop().execute(this::releaseQueuedBuffer);
+        }
+    }
+
+    @Override
+    public long getSessionId() {
+        return sessionId;
+    }
+
+    @Override
+    public int getSessionTimeout() {
+        return sessionTimeout;
+    }
+
+    @Override
+    public void process(WatchedEvent event) {
+        ReplyHeader h = new ReplyHeader(ClientCnxn.NOTIFICATION_XID, -1L, 0);
+        if (LOG.isTraceEnabled()) {
+            ZooTrace.logTraceMessage(
+                LOG,
+                ZooTrace.EVENT_DELIVERY_TRACE_MASK,
+                "Deliver event " + event + " to 0x" + Long.toHexString(this.sessionId) + " through " + this);
+        }
+
+        // Convert WatchedEvent to a type that can be sent over the wire
+        WatcherEvent e = event.getWrapper();
+
+        try {
+            int responseSize = sendResponse(h, e, "notification");
+            ServerMetrics.getMetrics().WATCH_BYTES.add(responseSize);
+        } catch (IOException e1) {
+            LOG.debug("Problem sending to {}", getRemoteSocketAddress(), e1);
+            close();
+        }
+    }
+
+    @Override
+    public int sendResponse(ReplyHeader h, Record r, String tag,
+                             String cacheKey, Stat stat, int opCode) throws IOException {
+        // cacheKey and stat are used in caching, which is not
+        // implemented here. Implementation example can be found in NIOServerCnxn.
+        if (closingChannel || !channel.isOpen()) {
+            return 0;
+        }
+        ByteBuffer[] bb = serialize(h, r, tag, cacheKey, stat, opCode);
+        int responseSize = bb[0].getInt();
+        bb[0].rewind();
```

Comment:
### ❌ New issue: Excess Number of Function Arguments
sendResponse has 6 arguments, threshold = 4